odoo/fields.py

```diff
@@ -3047,6 +3058,518 @@ def _record_ids_per_res_model(self, records):
+        """Return the value used in a read / search_read.
+
+        The value might contains the name_get of the many2one / many2many.
+        """
+        values = super().convert_to_read(value, record, use_name_get)
+        properties_definition = self._get_properties_definition(record) or []
+        if not values or not properties_definition:
+            return properties_definition
+
+        if isinstance(values, str):
+            values = json.loads(values)
+
+        if isinstance(values, dict):
+            values = self._dict_to_list(values, properties_definition)
+            values = self._parse_json_types(values, record.env)
+
+        assert isinstance(values, list), f"Wrong property type {type(values)!r}"
+
+        values = self._validate_values(values)
+        if use_name_get:
+            values = self._list_add_display_name(values, record.env)
+
+        return values
+
+    def convert_to_record(self, value, record):
+        return self.convert_to_read(value, record, use_name_get=False)
+
+    def convert_to_onchange(self, value, record, names):
+        # TODO: make the onchange work, not sure it's the right way to do it
+        # The record in argument has the old parent id, so when we call the super,
+        # convert_to_read is called, and the properties definition are fetched
+        # on the old parent, breaking the onchange
+        if not isinstance(value, list):
+            value = super().convert_to_onchange(value, record, names)
+
+        # add default values
+        for property_definition in value:
+            if property_definition.get('value') is None:
+                property_definition['value'] = property_definition.get('default')
+        return value
+
+    def _get_properties_definition(self, record):
+        parent = record[self.parent_record]
+        if not parent:
+            return
+
+        properties_definition = parent[self.parent_field]
+        return properties_definition
+
+    @classmethod
+    def _list_add_display_name(cls, values_list, env, value_key='value'):
+        """Add the name_get for each many2one / many2many properties.
+
+        :param values_list: List of properties definition and values
+        :param env: environment
+        """
+        for property_definition in values_list:
+            property_type = property_definition.get('type')
+            property_value = property_definition.get(value_key)
+            property_model = property_definition.get('model')
+            if not property_model:
+                continue
+
+            if property_type == 'many2one' and isinstance(property_value, int):
+                property_value = env[property_model].browse(property_value).name_get()
+                property_definition[value_key] = property_value[0] if property_value else False
+
+            elif property_type == 'many2many' and is_list_of(property_value, int):
+                property_value = env[property_model].browse(property_value).name_get()
```

Comment:
How does it behave if say the record (or one of the records in case of m2m) is filtered out by a record rule for a end user?

Does it just not show and will be erased if this same user saves the properties?
Would be nice to have a test for that.

Reply:
oops, forgot to answer this comment

Yes that's a tricky case, for now he will see an "empty record" (the record will be in the view but without name), so it won't get removed if the user save the form.

I don't think there's a good way to handle it without introducing security issue

I'll add a test